### DISCUSSION TITLE
Ability to tag tasks?

### DISCUSSION THREAD

**Discussion by p-hebert** [apache/airflow#17697] [Q&A] [answered]
I'm currently building a dag that self-populates its tasks based on the files (py, ipynb) of a directory.
I would like to access the xcom of all of those tasks once they are executed in order to retrieve their result from S3/GCS and then combine them in memory to assemble a finalized file.

Currently since these tasks are all of the same Operator type, I use the Operator type to filter the list of upstream tasks, but I plan to expand to different Operator types. While I could expand the list of Operator types that are filtered, I figured task tags would be cleaner.

I looked into the impl of BaseOperator, PythonOperator and some provider operators, and I couldn't find anything that looked like tags. The closest I found was the use of op_kwargs/op_args for PythonOperator, but not all operators have these parameters.
Hence:

1. Is there any way to tag operators? Alternatively I can monkey-patch tags on my operators, but that sounds hacky and dangerous.
2. Is this something that was considered in the past and decided against?
3. Any better solution?

**Comment by potiuk** [comment, marked as answer]
Tags are available at dag, not at task levels. But what I can suggest is a
naming convention for task_ids. That should do pretty well.

On Wed, Aug 18, 2021 at 9:15 PM Philippe Hebert ***@***.***>
wrote:

> I'm currently building a dag that self-populates its tasks based on the
> files (py, ipynb) of a directory.
> I would like to access the xcom of all of those tasks once they are
> executed in order to retrieve their result from S3/GCS and then combine
> them in memory to assemble a finalized file.
>
> Currently since these tasks are all of the same Operator type, I use the
> Operator type to filter the list of upstream tasks, but I plan to expand to
> different Operator types. While I could expand the list of Operator types
> that are filtered, I figured task tags would be cleaner.
>
> I looked into the impl of BaseOperator, PythonOperator and some provider
> operators, and I couldn't find anything that looked like tags. The closest
> I found was the use of op_kwargs/op_args for PythonOperator, but not all
> operators have these parameters.
> Hence:
>
>    1. Is there any way to tag operators? Alternatively I can monkey-patch
>    tags on my operators, but that sounds hacky and dangerous.
>    2. Is this something that was considered in the past and decided
>    against?
>    3. Any better solution?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/airflow/discussions/17697>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAERMI7CY6BUBUCC2KQLXPLT5QBDXANCNFSM5CMVVEUQ>
> .
>


-- 
+48 660 796 129

**Reply by p-hebert** [reply]
That is indeed functionally equivalent to a single tag functionality! :)
Alright, I'll use this as a solution! Thanks!

Were there any plans to implement tags at some point?/Are there any future plan to implement tags?

**Reply by potiuk** [reply]
Not that I am aware of  :)

**Reply by potiuk** [reply]
But if you would like to add it - feel free :)

**Reply by p-hebert** [reply]
:)
I won't commit, but I kinda want to add it 😄

**Reply by p-hebert** [reply]
@potiuk, after some extra hours going through the implementation for other stuff, I noticed that `BaseOperator.__init__` argument "params" seems to support usage as tags:

1. It's exposed as an attribute on the BaseOperator task:

https://github.com/apache/airflow/blob/c4d043d9f8b28f2ad80c9912e7690bc6620bd753/airflow/models/baseoperator.py#L617

2. It's directly read from the task copy in get_template_context

https://github.com/apache/airflow/blob/9922287a4f9f70b57635b04436ddc4cfca0e84d2/airflow/models/taskinstance.py#L1751-L1757

3. Bar overrides from other configurations, params is exposed in context:

https://github.com/apache/airflow/blob/9922287a4f9f70b57635b04436ddc4cfca0e84d2/airflow/models/taskinstance.py#L1975

Hence if I understood correctly:

1. I could call the operator with `params={"task_tags": ("tag:one", "tag:two")}` and be able to access those tags directly in a templated string
2. I could instanciate the other operators with params as in (1) and then filter them through their params since it is an attribute on the task.

Hence I think tagging is actually already supported :)

**Reply by potiuk** [reply]
If you can make it works for you - that's cool. I think the para,s are not in any way addressed in UI, but if it's not needed for you that might do the job!

**Reply by p-hebert** [reply]
I confirm, it does :)

**Reply by potiuk** [reply]
Cool. I love new features, that reqiuire 0 new lines of code to maintain 😃

**Reply by jose-padron-momox** [reply]
i would like to have that in v3... sometimes is not enough to have task names templated, a nice tag/label helps too
